benchmarks/solubility/plugin/solubility.cc

```diff
@@ -0,0 +1,409 @@
+         * if and only if base model is compressible.
+         */
+        bool is_compressible () const override;
+
+        /**
+         * Compute the equilibrium melt fractions for the given input conditions.
+         * @p in and @p melt_fractions need to have the same size.
+         *
+         * @param in Object that contains the current conditions.
+         * @param melt_fractions Vector of doubles that is filled with the
+         * equilibrium melt fraction for each given input conditions.
+         */
+        virtual void melt_fractions (const MaterialModel::MaterialModelInputs<dim> &in,
+                                     std::vector<double> &melt_fractions) const;
+
+        virtual double reference_darcy_coefficient () const;
+
+        /**
+         * Method to declare parameters related to volatile model
+         */
+        static void
+        declare_parameters (ParameterHandler &prm);
+
+        /**
+         * Method to parse parameters related to volatile model
+         */
+        void
+        parse_parameters (ParameterHandler &prm) override;
+
+        virtual
+        void
+        create_additional_named_outputs (MaterialModel::MaterialModelOutputs<dim> &out) const;
+
+      private:
+        /**
+         * Pointer to the material model used as the base model
+         */
+        std::unique_ptr<MaterialModel::Interface<dim>> base_model;
+
+        double reference_rho_f;
+        double shear_to_bulk_viscosity_ratio;
+        double eta_f;
+        double reference_permeability;
+        double alpha_phi;
+        double melt_compressibility;
+        double melting_time_scale;
+    };
+
+
+
+    template <int dim>
+    void
+    Volatiles<dim>::initialize()
+    {
+      base_model->initialize();
+    }
+
+
+
+    template <int dim>
+    void
+    Volatiles<dim>::update()
+    {
+      base_model->update();
+    }
+
+
+
+    template <int dim>
+    void
+    Volatiles<dim>::evaluate(const typename Interface<dim>::MaterialModelInputs &in,
+                             typename Interface<dim>::MaterialModelOutputs &out) const
+    {
+      base_model->evaluate(in,out);
+
+      // Modify the viscosity from the base model based on the presence of melt.
+      const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
+
+      if (in.requests_property(MaterialProperties::viscosity))
+        {
+          // Scale the base model viscosity value based on the porosity.
+          for (unsigned int q=0; q<out.n_evaluation_points(); ++q)
+            {
+              const double porosity = std::max(in.composition[q][porosity_idx],0.0);
+              out.viscosities[q] *= (1.0 - porosity) * exp(- alpha_phi * porosity);
+            }
+        }
+
+      // Fill the melt outputs if they exist.
+      MeltOutputs<dim> *melt_out = out.template get_additional_output<MeltOutputs<dim>>();
+
+      if (melt_out != nullptr)
```

Comment:
Could this section of code be expanded to have a condition where the MeltOutputs are not filled, and instead the fluids are advected using the single-phase Darcy method? 

As an alternative, could a trick be done where the porosity given to the MeltOutputs is set to 0, such that is solves the normal equations for solid deformation and the fluids are advected with the darcy field?

Reply:
I think you could just switch off melt transport and then change the porosity to be of type darcy_field. I do not think this would require changing the code, but there may be some things I am not considering at the moment.